### PR TITLE
Add Internal.AspNetCore.SiteExtension.Sdk package

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/Universe/pull/697

### CODE REVIEW

File: src/Internal.AspNetCore.HostingStartup.Sdk/TrimDeps.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public class TrimDeps : Task
+    {
+        [Required]
+        public ITaskItem[] DepsFiles { get; set; }
+
+        public override bool Execute()
+        {
+            foreach (var depsFile in DepsFiles)
+            {
+                ChangeEntryPointLibraryName(depsFile.GetMetadata("Identity"));
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/TrimDeps.cs]: FYI - this is identical to `depsFile.ItemSpec`. I have a slight preference for this instead of .GetMetadata, but its functionally the same.

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.props [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+-->
+<Project TreatAsLocalProperty="TaskFolder;TaskAssembly">
+  <PropertyGroup>
+    <TaskFolder Condition=" '$(MSBuildRuntimeType)' == 'Core' ">netstandard1.6</TaskFolder>
+    <TaskFolder Condition=" '$(MSBuildRuntimeType)' != 'Core' ">net46</TaskFolder>
+    <TaskAssembly>$(MSBuildThisFileDirectory)..\tasks\$(TaskFolder)\Internal.AspNetCore.HostingStartup.Sdk.dll</TaskAssembly>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Remove="obj\depswork\*.*" />
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.props]: BaseIntermediateOutputPath is excluded by default. Why was this needed?

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  This file is automatically imported by NuGet into a user's project
+  when it targets a single framework, or in classic (pre 2017) csproj projects.
+-->
+
+<Project>
+
+  <Target Name="GenerateHostingStartupDeps" BeforeTargets="Build" Condition="'@(HostingStartupPackageReference->Count())' != '0'">
+    <PropertyGroup>
+      <_TemplatesDirectory>$(MSBuildThisFileDirectory)..\content\</_TemplatesDirectory>
+      <_DepsOutputDirectory>$(MSBuildProjectDirectory)\$(BaseIntermediateOutputPath)</_DepsOutputDirectory>
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: Use `$(IntermediateOutputPath)` instead of BaseIntermediateOutputPath. This will already have been normalized for config + target framework, and have been turned into a full path.

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      <_WorkingDirectory>$(_DepsOutputDirectory)\depswork</_WorkingDirectory>
+      <_BasePackagePath>content\additionaldeps\</_BasePackagePath>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_TemplateFiles Include="$(MSBuildThisFileDirectory)..\content\HostingStartup\*.*" />
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: Should this be recursive? This will only capture files in the top directory.

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    </ItemGroup>
+
+    <RemoveDir Directories="%(_HostingStartupPackageReference.WorkingDirectory)" />
+
+    <Copy SourceFiles="@(_TemplateFiles)" DestinationFolder="%(_HostingStartupPackageReference.WorkingDirectory)" />
+
+    <MSBuild Projects="%(_HostingStartupPackageReference.Project)" Targets="Restore" />
+
+    <MSBuild Projects="%(_HostingStartupPackageReference.Project)"
+             Targets="Restore;Clean;Publish"
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: Remove duplicate call to the Restore target

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        TrimmedDepsFile="$(_DepsOutputDirectory)%(HostingStartupPackageReference.Identity)\%(HostingStartupPackageReference.Identity).deps.json"
+        PackagePath="%(HostingStartupPackageReference.Identity)\shared\Microsoft.NETCore.App\$(HostingStartupRuntimeFrameworkVersion)\"
+         />
+    </ItemGroup>
+
+    <RemoveDir Directories="%(_HostingStartupPackageReference.WorkingDirectory)" />
+
+    <Copy SourceFiles="@(_TemplateFiles)" DestinationFolder="%(_HostingStartupPackageReference.WorkingDirectory)" />
+
+    <MSBuild Projects="%(_HostingStartupPackageReference.Project)" Targets="Restore" />
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: Restore shouldn't succeed unless you have the properties set `HostingStartupPackageName=%(_HostingStartupPackageReference.Identity);HostingStartupPackageVersion=%(_HostingStartupPackageReference.Version)`, right?

File: src/Internal.AspNetCore.HostingStartup.Sdk/buildMultiTargeting/Internal.AspNetCore.HostingStartup.Sdk.props [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<!--
+  This file is automatically imported by NuGet into a user's project
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/buildMultiTargeting/Internal.AspNetCore.HostingStartup.Sdk.props]: What are the scenarios for mutli-targeting hosting startup projects? Adding these imports is unlikely to work in a multi-target scenario since these projects invoke a new project build context per TFM.

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    <MSBuild Projects="%(_HostingStartupPackageReference.Project)"
+             Targets="Restore;Clean;Publish"
+             Properties="PublishDir=%(_HostingStartupPackageReference.WorkingDirectory);HostingStartupPackageName=%(_HostingStartupPackageReference.Identity);HostingStartupPackageVersion=%(_HostingStartupPackageReference.Version)" />
+
+    <Copy SourceFiles="%(_HostingStartupPackageReference.DepsFile)" DestinationFiles="%(_HostingStartupPackageReference.TrimmedDepsFile)" />
+
+    <RepoTasks.TrimDeps DepsFiles="%(_HostingStartupPackageReference.TrimmedDepsFile)" />
+
+    <ItemGroup>
+      <Content Include="%(_HostingStartupPackageReference.TrimmedDepsFile)" PackagePath="%(_HostingStartupPackageReference.PackagePath)" />
```
Comment (natemcmaster) [inline src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: May not work when calling `dotnet pack --no-build` and in multi-targeting scenarios. Targets that add custom content to packages should add items to the `BuildOutputInPackage` item group and chain into TargetsForTfmSpecificBuildOutput.

cref https://github.com/NuGet/NuGet.Client/pull/1255

Example:
```xml
<PropertyGroup>
  <TargetsForTfmSpecificBuildOutput>$(TargetsForTfmSpecificBuildOutput);MyGenTarget</TargetsForTfmSpecificBuildOutput>
</PropertyGroup>

<Target Name="MyGenTarget">
 <!-- do stuff-->
  <ItemGroup>
     <BuildOutputInPackage Include="myfile.txt" />
   </ItemGroup>
</Target>
```

File: src/Internal.AspNetCore.HostingStartup.Sdk/buildMultiTargeting/Internal.AspNetCore.HostingStartup.Sdk.props (hunk 7 above)
Comment (pakrym) [reply src/Internal.AspNetCore.HostingStartup.Sdk/buildMultiTargeting/Internal.AspNetCore.HostingStartup.Sdk.props]: I just copied this from the sample. Would it work fine if I remove them?

Comment (natemcmaster) [reply src/Internal.AspNetCore.HostingStartup.Sdk/buildMultiTargeting/Internal.AspNetCore.HostingStartup.Sdk.props]: Yup, it should work ok

File: src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets (hunk 8 above)
Comment (pakrym) [reply src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: Problem with `BuildOutputInPackage` is that it put's files into `lib` subdirectory even if `TargetPath` is set. So I have to set it to `TargetPath="..\..\$(_BasePackagePath)"`

Comment (pakrym) [reply src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: I think I'll stick with `Content` for now and see if we have any issues.

Comment (natemcmaster) [reply src/Internal.AspNetCore.HostingStartup.Sdk/build/Internal.AspNetCore.HostingStartup.Sdk.targets]: Hmm, what about <TfmSpecificPackageFile> + TargetsForTfmSpecificContentInPackage?

https://github.com/NuGet/NuGet.Client/blob/4.3.0.4339/src/NuGet.Core/NuGet.Build.Tasks.Pack/Pack.targets#L275-L278

Also, you might consider simplifying this by just using a nuspec file. I've found this is the only way to do anything custom.

cc @rohit21agrawal

File: src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    <Description>Build targets and extensions to Microsoft.NET.Sdk. This package is intended for Microsoft use only.</Description>
+    <NoPackageAnalysis>true</NoPackageAnalysis>
+    <NuspecFile>$(MSBuildThisFileDirectory)$(MSBuildProjectName).nuspec</NuspecFile>
+    <IncludeSymbols>false</IncludeSymbols>
+    <IncludeSource>false</IncludeSource>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="$(Tooling_NewtonsoftJsonPackageVersion)" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="15.1.1012" />
```
Comment (JunTaoLuo) [inline src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: Are we sure we want these as hardcoded version numbers? @natemcmaster

File: src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj [hunk 10]
```diff
@@ -0,0 +1,32 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <PackageId>Internal.AspNetCore.SiteExtension.Sdk</PackageId>
+    <TargetFrameworks>netcoreapp2.0;net46</TargetFrameworks>
```
Comment (JunTaoLuo) [inline src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: Just double checking to see we want to target netcoreapp2.0 and not 2.1?

Comment (JunTaoLuo) [reply src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: Actually, this is an internal tool so this makes sense.

File: src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj (hunk 9 above)
Comment (natemcmaster) [reply src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: True. There are package versions available for this. Would be better to use those.

File: src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj (hunk 10 above)
Comment (natemcmaster) [reply src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: Right. This is for a task assembly, not a runtime one.

File: src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj (hunk 9 above)
Comment (pakrym) [reply src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: fixed

File: src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj (hunk 10 above)
Comment (pakrym) [reply src/Internal.AspNetCore.SiteExtension.Sdk/Internal.AspNetCore.SiteExtension.Sdk.csproj]: Same as in `Internal.AspNetCore.Sdk`

**Comment by pakrym** [comment]
@JunTaoLuo 
Site extension project would look the following way:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <HostingStartupRuntimeFrameworkVersion>2.0.0</HostingStartupRuntimeFrameworkVersion>
  </PropertyGroup>

  <ItemGroup>
    <HostingStartupPackageReference Include="microsoft.aspnetcore.azureappservices.hostingstartup" Version="2.0.1" />
    <HostingStartupPackageReference Include="microsoft.aspnetcore.applicationinsights.hostingstartup" Version="2.0.1" />

    <PackageReference Include="Internal.AspNetCore.SiteExtension.Sdk" Version="2.1.0-*" />
  </ItemGroup>

</Project>

```

And when you dotnet pack it magic things would happen and deps and dlls would get embedded into the package.
